### ISSUE TITLE
TypeError: Cannot use 'in' operator to search for 'onfocusin' in undefined

### ISSUE THREAD

**Issue by nullhook** [carbon-design-system/carbon#442]
I'm getting an error while importing carbon components like below:

`import {Modal, DataTableV2} from 'carbon-components';`

`TypeError: Cannot use 'in' operator to search for 'onfocusin' in undefined`
` at Function.init (/node_modules/carbon-components/umd/globals/js/mixins/init-component-by-event.js:61:116)`

`TypeError: target.querySelectorAll is not a function`
` at Function.init (/node_modules/carbon-components/umd/globals/js/mixins/init-component-by-event.js:61:116)`

Has anyone here faced this issue while importing? Do I need to load any dependencies with this?

**Comment by asudoh** [comment]
@rehat101 Do you have a public website or a like that demonstrates the issue you faced? If you are an IBMer another option is putting your code (that can easily run) to our GHE and dropping a link to our Slack. Thanks!

**Comment by nullhook** [comment]
@asudoh - I have created a sample project here https://github.com/rehat101/carbon — you can run this: `donejs develop`

**Comment by asudoh** [comment]
Thanks @rehat101 - Seems that your app uses CanJS and according to https://github.com/canjs/can-simple-dom/issues/25 it lacks [`HTMLDocument.defaultView`](https://developer.mozilla.org/en-US/docs/Web/API/Document/defaultView). Carbon uses that property to get the `window` object associated with the DOM element Carbon component is created on. A workaround for such CanJS issue would be running `document.defaultView = window` code before `DOMContentLoaded` event fires.

**Comment by nullhook** [comment]
@asudoh - After adding `global.document.defaultView = global` this here: https://github.com/canjs/can-vdom/blob/master/make-window/make-window.js#L422 - I'm getting a new type of error now.

`TypeError: target.querySelectorAll is not a function` 

I believe Carbon needs more APIs than what CanJS DOM provides?

**Comment by nullhook** [comment]
I also tried doing this:

```
import {Modal} from 'carbon-components';
import isNode from 'detect-node';

if(!isNode) {
  Modal.init();
}
```

and I get the same error above 
`TypeError: target.querySelectorAll is not a function`

**Comment by asudoh** [comment]
@rehat101 As the error implies, CanJS doesn’t seem to support [qSA](https://developer.mozilla.org/en-US/docs/Web/API/Document/querySelectorAll). I personally was surprised by CanJS’ lack of support for that API because it’s a quite old and very frequently used API.

Given Carbon vanilla JS code doesn’t do rendering (it basically hooks event handlers to DOM elements and mangles DOM in event handler code), your attempt to avoid running Carbon code on server makes all sense to me. To do that, you need to disable Carbon’s “automatic instantiation upon `DOMContentLoaded`” feature (mentioned in “Don't initialize components by default” section in the [Getting Started doc](http://www.carbondesignsystem.com/getting-started/developers)), by:

```javascript
import * as CarbonComponents from 'carbon-components';
import isNode from 'detect-node';

if (isNode) {
  CarbonComponents.settings.disableAutoInit = true;
}
```

**Comment by nullhook** [comment]
@asudoh - Ok, seems like that error is disabled now. Where and how do I instantiate components now?

**Comment by asudoh** [comment]
@rehat101 The API to instantiate a Carbon components is `CarbonComponents.ComponentClass.create()`, e.g.:

```javascript
CarbonComponents.OverflowMenu.create(document.querySelector('[data-overflow-menu]'));
```

Carbon runs it automatically upon `DOMContentLoaded` event, but many JavaScript frameworks constructs the corresponding DOM (e.g. element with `data-overflow-menu` attribute) later than that, and it seems to be the case for your CanJS application, too.

"Wrapping a component with JavaScript framework of your choice" section in http://www.carbondesignsystem.com/getting-started/developers discusses this topic further, and https://codepen.io/asudoh/pen/VryJBO as well as https://codepen.io/asudoh/pen/OOOMdB provides some examples for this topic.

In a nutshell, you'll need to see if CanJS has a lifecycle event/callback for component DOM being ready, something similar to [Angular2+ `ngAfterViewInit()`](https://angular.io/api/core/AfterViewInit), [Angular1 `link()`](https://docs.angularjs.org/guide/directive#creating-a-directive-that-manipulates-the-dom) or [Web Components'](https://developer.mozilla.org/en-US/docs/Web/Web_Components/Custom_Elements) `connectedCallback()`. If there is, it's as easy as calling `CarbonComponents.ComponentClass.create()` API upon such event. If not, you'll need to poll for DOM modification that would slow done your app. `CarbonComponents.watch()` API does such polling for you in the fastest possible way (with DOM [`MutationObserver` API](https://developer.mozilla.org/en-US/docs/Web/API/MutationObserver)).

**Comment by nullhook** [comment]
@asudoh - so I figured CanJS's lifecycle events and running the `create()` method at an app level because my component is at an app level.

```
AppViewModel.prototype._eventSetup = function(){
  CarbonComponents.OverflowMenu.create(document.querySelector('[data-overflow-menu]'));
};
```

Here: https://github.com/rehat101/carbon/blob/master/src/app.js#L26

**Comment by asudoh** [comment]
@rehat101 Seems that DOM creation happens *after* `_eventSetup()` - Putting a `debugger` statement in that function and setting DOM node removal breakpoint in Chrome DevTools for the DOM element with `data-overflow-menu` attribute when `_eventSetup()` runs, I see the following code running after then, which seems in the compiled version of [your `index.stache` file](https://github.com/rehat101/carbon/blob/master/src/index.stache):

```javascript
/**
 * @function attach
 * @hide
 * @description Receives the completely rendered DocumentFragment and
 * attaches the parts from the head into the document.head, the body into
 * document.body.
 * @signature `attach(result)`
 * @param {RenderResult} The result of rendering within a Zone.
 */
function attach(result){
  var frag = result.fragment;

  // If already attached skip this part.
  if(document.documentElement.hasAttribute("data-attached")) {
    return;
  }

  var head = document.head;
  var body = document.body;

  // Move elements from the fragment's head to the document head.
  eachChild(head, remove);

  var fragHead = frag.querySelector("head");
  eachChild(fragHead, appendTo(head));

  // Move elements from the fragment's body to the document body.
  eachChild(body, remove);

  var fragBody = frag.querySelector("body");
  eachChild(fragBody, appendTo(body));
  document.documentElement.setAttribute("data-attached", "");
}
```

**Comment by nullhook** [comment]
@asudoh I have tried the following way and it seems to be working however I don't know if this is the proper way or If i'm breaking CanJS rules.

What are your thoughts? 

https://github.com/rehat101/carbon/blob/master/src/app.js#L12

Edit: Actually, this is not the best solution because sometimes it doesn't find the document and I end up getting dqs error

**Comment by asudoh** [comment]
@rehat101 Honestly I hadn’t tried CanJS before this issue came along and thus don’t have many words on CanJS rules, but the best practice to use a JavaScript framework in general with Carbon vanilla code is creating “wrapping components”, described in the corresponding section in: http://www.carbondesignsystem.com/getting-started/developers

Most JavaScript frameworks provide component model with lifecycle events/callbacks, and it seems the case for CanJS, too. Here’s an example: https://codepen.io/asudoh/pen/eeKzOq

**Comment by nullhook** [comment]
@asudoh - What if I have 3 different overflow menus with different menu options does that mean I need 3 different "wrapping components" ?

**Comment by asudoh** [comment]
@rehat101 If you take a look at my example, you have two "wrapping components" (classes), one is `<bx-overflow-menu>` and another is `<bx-overflow-menu-option>`. Note that you don't have to create those classes (by `can.Component.extend()`) for every instance. For example, for the following CanJS template in my example, `can.Component.extend({ tag: 'bx-overflow-menu-option', ... })` is called only once:

```html
<script type='text/stache' id='app'>
  <bx-overflow-menu>
    <bx-overflow-menu-option>Foo</bx-overflow-menu-option>
    <bx-overflow-menu-option>Bar</bx-overflow-menu-option>
    <bx-overflow-menu-option>Baz</bx-overflow-menu-option>
  </bx-overflow-menu>
</script>
```

Hope it addresses your question.

**Comment by nullhook** [comment]
@asudoh - makes sense but I feel like the "wrapping components" isn't feasible perhaps because I use all Carbon components and this means I'll need 10-15 wrapping components. 

Is there an alternative way where I can initialize all components without "wrapping" them or breaking any DoneJS rules too?

**Comment by asudoh** [comment]
@rehat101 You might want to try creating a higher-order wrapping component that contains bigger portions of your application e.g. https://github.com/rehat101/carbon/blob/6d6a98a/src/index.stache#L12-L60 in its template. Then you can instantiate all Carbon components in that template in `can.Component#init()`. Hope this helps.

**Comment by nullhook** [comment]
@asudoh Thanks for the suggestion. I'll try that and let you know.

I'm currently facing another issue with alignment of menu options which overlaps with the browser window:
![image](https://user-images.githubusercontent.com/8665427/33512915-a3cb8e7e-d75e-11e7-92bf-617c156eb517.png)

**Comment by asudoh** [comment]
@rehat101 Great - Wrt the truncated overflow menu in above screenshot, you might want to try adding `bx--overflow-menu—flip` class to `<ul class="bx--overflow-menu-options”>`.

**Comment by nullhook** [comment]
@asahlin - I still get a truncated overflow menu when adding `bx--overflow-menu—flip`

**Comment by asudoh** [comment]
@rehat101 Did you get right-aligned overflow menu (instead of left-aligned one)? If not, please be advised that `bx--overflow-menu—flip` class should be applied to `<ul class="bx--overflow-menu-options”>` instead of `<div data-overflow-menu>`. Thanks!

**Comment by nullhook** [comment]
@asudoh as you can see in the inspector shot below the class is applied however the menu still truncates

![image](https://user-images.githubusercontent.com/8665427/33539645-802c6566-d8ed-11e7-8dd1-c8e11eac902e.png)

**Comment by asudoh** [comment]
@rehat101 Do you mean you still have left-aligned overflow menu?

**Comment by nullhook** [comment]
@asudoh correct but it's not really left aligned. Any instance of overflow menu half of the UI hides.

**Comment by asudoh** [comment]
@rehat101 Do you have a screenshot of the truncated overflow menu after you made it right-aligned? Thanks!

**Comment by nullhook** [comment]
@asudoh here you go

![image](https://user-images.githubusercontent.com/8665427/33541896-e9a99dc0-d8f6-11e7-9b50-b80d945536e3.png)

**Comment by asudoh** [comment]
@rehat101 You seem to still have left-aligned overflow menu. Given it’s no longer relevant to the error this issue originally describes, you may create another issue with a public site link that reproduces the flip style being ignored. Thanks!
